### ISSUE TITLE
Refutational soundness bug on QF_BVFP instance

### ISSUE THREAD

**Issue by fwangdo** [bitwuzla/bitwuzla#130]
Greetings,
For this instance, a refutational soundness bug occurred.
(we checked that the instance had a solution through z3.)
```
$ ./bitwuzla ./small.smt2
unsat
[error] ./small.smt2:12:11: model generation is not enabled

$ z3 ./small.smt2 model_validate=true
sat
(
  (define-fun generated () (_ FloatingPoint 10 47)
    (_ NaN 10 47))
  (define-fun x_ackermann!3 () (_ BitVec 64)
    #xfff0000000000000)
  (define-fun x_ackermann!0 () (_ BitVec 64)
    #x000fffffffffffff)
  (define-fun x_ackermann!2 () (_ BitVec 64)
    #x0000000000000000)
  (define-fun x_ackermann!5 () (_ BitVec 64)
    #x0000000000000000)
  (define-fun x_ackermann!1 () (_ BitVec 64)
    #x0000000000000000)
  (define-fun x_ackermann!4 () (_ BitVec 64)
    #x0000000000000000)
)

$ cat ./small.smt2
(set-logic QF_BVFP)
(declare-fun x_ackermann!1 () (_ BitVec 64))
(declare-fun x_ackermann!0 () (_ BitVec 64))
(declare-fun x_ackermann!2 () (_ BitVec 64))
(declare-fun x_ackermann!3 () (_ BitVec 64))
(declare-fun x_ackermann!4 () (_ BitVec 64))
(declare-fun x_ackermann!5 () (_ BitVec 64))
(declare-fun generated () (_ FloatingPoint 10 47))
(assert (not (bvule ((_ fp.to_sbv 16) RNE (fp.div RNA ((_ to_fp 4 12) RTP (bvor #b0100001000110110 #b1001110010000011)) ((_ to_fp 4 12) RNE ((_ sign_extend 10) #b101010)))) (bvurem ((_ fp.to_sbv 16) RTZ ((_ to_fp_unsigned 4 12) RNE (bvurem #b0101100101001000 #b1111001000010000))) ((_ zero_extend 6) ((_ fp.to_sbv 10) RNA ((_ to_fp_unsigned 5 5) RTN #b1000110111)))))))
(assert (not (and (not (fp.isNaN ((_ to_fp 11 53) x_ackermann!0))) (or (or (fp.isNaN ((_ to_fp 11 53) (bvxor ((_ sign_extend 7) (bvand (bvashr ((_ fp.to_sbv 57) RTZ generated) (bvurem #b011010100111010101011010101000101000010111110001100111101 #b000000011010111110110111110111111001100110011111111100011)) (bvmul (bvsdiv #b110001100111110111101110000011010111011000011001101111100 #b111100000111011101101011010101111000010100011010010100010) ((_ fp.to_sbv 57) RTP (fp #b0 #b10000000000000000000110111 #b110011110011000000101000001001))))) (bvshl (bvnot (bvshl ((_ fp.to_sbv 64) RNE (fp #b0 #x83e #b101011010000110001110010010100011100010101011001000)) (bvadd #b0011111100100110000010011001100010001101111100000100110010000111 #b1111100010011101001111100100101101010111001010100001101101010110))) ((_ extract 78 15) (bvlshr (bvshl #b11100010101111111110110000111001110010011100100010000010010110000001111101000000110100100000100111100001111110001000001010000001 #b01010011010000000011110000110001001010100111110001111011110010011011110100000110011101101011110110001110001100011100000011001100) (bvor #b01011001111110100101000111100000101100101000001100010111001000101000010011001111110111111100111100101101011011011000111010100101 #b10001101110100000111000110100101010001100001110111110110011110011001011010011100111101111000111100000110000101011111001111110010))))))) (fp.isNaN ((_ to_fp 11 53) x_ackermann!0))) (fp.gt ((_ to_fp 11 53) x_ackermann!3) ((_ to_fp 11 53) x_ackermann!0))))))
(check-sat)
(get-model)
(exit)
```
commit version: [2b3d69a](https://github.com/bitwuzla/bitwuzla/commit/2b3d69a797e5c60f83a9f8fc200f03fadcf7f4ad)
release version:

**Comment by BrunoDutertre** [comment]
This bug can be reduced to this:
```
(set-logic QF_BVFP)
(assert (= (fp.div RNA (fp #b1 #xe #b11111111111) (fp #b1 #xb #b01100000000))
	   (fp #b0 #xa #b01110100010)))
(check-sat)
```

Both z3 and cvc5 say that this is sat. Bitwuzla says that this is unsat.  In debug mode, bitwuzla crashes:
```
Assertion failed: (bw <= op.getWidth()), function matchWidth, file symfpu_wrapper.cpp, line 420.
```

**Comment by BrunoDutertre** [comment]
It may also depend on the `symfpu` version bitwuzla is built with.

**Comment by mpreiner** [comment]
cvc5 fails with a similar assertion failure in debug mode.

**Comment by martin-cs** [comment]
I have successfully reproduced these and will see what I can do.

**Comment by aniemetz** [comment]
I dug into why we answer unsat when cvc5 answers sat. Using Bruno's example here, because it's simpler.

This example only requires constant folding, which is done via SymFPU for both solvers. However, cvc5 correctly
answering sat is purely **by accident**.

What happens is that in `symfpu::arithmeticDivide`, SymFPU computes a format for `divideResult` that triggers an
extension operation via `matchWidth` where the number of bits `n` to extend with is `-1` (due to `bw > op.getWidth()`,
hence the assertion failure). And since `n` is an `uint32_t`, this is `2 **32 - 1`. Now, in cvc5, we compute the size of the
resulting bit-vector as `d_size + n`, which causes an overflow as `d_size` is an uint32_t, and thus, by luck, it arrives at
a sane bit-width and the correct result. However, in Bitwuzla's bit-vector implementation, the size of a bit-vector is a
`uint64_t`, hence no overflow, hence not a sane bit-width but a very very large bit-width, which in the end causes the
constant fold to produce `+zero` instead of the correct result.

Bottom line, this is entirely a SymFPU problem, and not a problem on our side. For now, we have disabled experimental
formats on main via a compile time option by default. We consider any format that is not `Float16`, `Float32`, `Float64`
or `Float128` as experimental. Note that cvc5 has similar guards in place, however, configurable via a runtime option,
and it fails to guard the constant-folding case (which is an oversight and will be fixed).

**Comment by aniemetz** [comment]
We now use MPFR for constant folding and this specific test case is now fixed on main. The underlying issue in the SymFPU encoding has been fixed on an experimental branch of SymFPU. We will switch to use that fixed branch soon.
